### PR TITLE
[serve] Add health checking for http proxy actors

### LINKED ISSUE #19151: Properly health check HTTP proxy & report its status
One idea is to make the HTTP proxy a "normal" deployment. The behavior we want w/ the HTTP proxies is quite similar to other deployments and this would allow us to deduplicate code and dogfood deployments a bit. We could also get some benefits such as the ability to do rolling updates for config and argument changes (e.g., route table updates, changing the port the HTTP server is listening on or adding a middleware).

The only thing blocking this at the moment is the ability to have resizable placement groups, which would allow us to maintain the existing placement strategy of one proxy per node. There may be a hacky temporary workaround for this, but I haven't thought of a good one yet (I'd like to avoid hard-coding scheduling logic for this special case).

Prototype: https://github.com/ray-project/ray/pull/15820

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Add health checking to HTTP Proxy
* Add no-op `check_health()` method to proxy actor
* In controller, periodically call `check_health()` method
  * Before ready() returns, status is STARTING
  * If an error occurs with check_health(), set UNHEALTHY
  * If request to check_health() times out, set UNHEALTHY

## Related issue number

Closes #19151

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by sihanwang41 — COMMENTED** [review]
Nice work! leave some nits and please make sure unit tests cover all the statues :)

File: python/ray/serve/_private/http_state.py [hunk 1]
```diff
@@ -20,11 +21,75 @@
...
+
+        if self._health_check_obj_ref:
+            finished, _ = ray.wait([self._health_check_obj_ref], timeout=0)
+            if finished:
+                self._check_health_obj_ref_result()
+        # If there's no active in-progress health check and it has been more than 10
+        # seconds since the last health check, perform another health check
+        if time.time() - self._last_health_check_time > randomized_period_s:
+            # If the HTTP Proxy has been blocked for more than 5 seconds, mark unhealthy
+            if time.time() - self._last_health_check_time > 5:
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/http_state.py]: Changed!

File: python/ray/serve/_private/http_state.py [hunk 2]
```diff
@@ -20,11 +21,75 @@
...
     get_all_node_ids,
 )
-from ray.serve._private.common import EndpointTag, NodeId
+from ray.serve._private.common import EndpointTag, NodeId, HTTPProxyStatus
+from ray.serve.schema import HTTPProxyDetails
 
 logger = logging.getLogger(SERVE_LOGGER_NAME)
 
 
+PROXY_HEALTH_CHECK_PERIOD_S = 10
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/http_state.py]: Moved to constants.py

File: python/ray/serve/_private/http_proxy.py [hunk 3]
```diff
@@ -549,3 +549,6 @@ async def run(self):
 
         self.setup_complete.set()
         await server.serve(sockets=[sock])
+
+    async def check_health(self):
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/http_proxy.py:553]: Added!

**Comment by edoakes** [comment]
`If request times out, set HEALTHY` <- did you mean `UNHEALTHY` ?

**Comment by zcin** [comment]
> `If request times out, set HEALTHY` <- did you mean `UNHEALTHY` ?

Oops yes, updated

File: python/ray/serve/schema.py [hunk 4]
```diff
@@ -758,6 +759,11 @@ def get_status_dict(self) -> Dict:
         )
 
 
+@PublicAPI(stability="alpha")
+class HTTPProxyDetails(BaseModel):
+    status: HTTPProxyStatus
```
Comment (edoakes) [inline python/ray/serve/schema.py:764]: can we add node_id, ip_address here?

@alanwguo anything else that there would be useful to show in the UI?

oh... maybe logs path?

Comment (zcin) [reply python/ray/serve/schema.py]: I've added metadata for http proxy here: https://github.com/ray-project/ray/pull/34944/files#diff-e454e14e8c39af64fcec6997326bd32012fc91afe5d09d6ac1486aed6199afd1R764-R776

**Review by edoakes — APPROVED** [review]
This is awesome, thanks for picking it up!

**Pull request merged**